### DISCUSSION TITLE
Restructuring Packages & Directories

### DISCUSSION THREAD

**Discussion by danielbate** [FuelLabs/fuels-ts#1517] [Ideas]
We currently have 26 packages in the `packages` directory. This is an unnecessary level of abstraction that increases `build`, `watch`, `install` and `test` times. We should look to restructure `packages` into a smaller logical structure. Further to this, some that were acting as a bridge between packages may be deprecated when relocated if this now makes logical sense (i.e. `@fuel-ts/testcases`). This issue will act as a topic for all restructuring activities, and further issues should be created per package.

**Proposed Structure**

<img width="864" alt="Screenshot 2023-12-07 at 17 46 01" src="https://github.com/FuelLabs/fuels-ts/assets/9402987/4393c04c-d0f3-47df-9790-40b22484db7f">

**Comment by arboleya** [comment]
@danielbate Thanks for creating the issue.

Should we have individual issues per group, though?

**Comment by danielbate** [comment]
@arboleya agreed. I had added the following:

> This issue will act as a topic for all restructuring activities.

But will make it more explicit around further issues.

**Comment by petertonysmith94** [comment]
Just took another quick look over this one (for my understanding of the dependencies referenced in #1356).

The consolidation of the `account` package has been pretty big IMO (thanks @Torres-ssf 🥇). I would say the proposed `lib` package would be a massive win.

<img width="1307" alt="image" src="https://github.com/FuelLabs/fuels-ts/assets/16990131/41cfe9a9-c0f2-42de-8a05-84984d1404d1">

[Generated using `pnpm-workspace-graph`](https://classic.yarnpkg.com/en/package/pnpm-workspace-graph)

**Comment by maschad** [comment]
@danielbate is there a reason why the `interfaces` , `merkle` and `hasher` libraries are unattached? I also wonder if `math` , `address`, `crypto`, `hasher` , `merkle` and `transaction` should be grouped together in `lib`, then `test-utils` , `error`. `versions` and `interfaces` should be in `utils` ?

**Reply by arboleya** [reply]
@maschad Thanks for your observations!

They should be _attached_; I probably messed up and deleted their lines.

The packages `math`, `address`, `crypto`, `hasher`, and `merkle` — _are_ under `lib` tho.

Only `transaction` is not, but it could be, and we could also include a `utils` package.

Grouping these is always subjective. We must carefully think it through.

**Reply by maschad** [reply]
Thanks for the clarification. Based on the dependency graph @petertonysmith94 generated it seems that suggestion could work.

**Comment by petertonysmith94** [comment]
This discussion would be closed via:
- https://github.com/FuelLabs/fuels-ts/issues/2493

**Reply by maschad** [reply]
I think we can conclude the discussion with the issue outlined in https://github.com/FuelLabs/fuels-ts/issues/2493

We also arrived at consensus around this at the offsite.

**Comment by petertonysmith94** [comment]
Closing this discussion, we're moving to a single `fuels` package:

- https://github.com/FuelLabs/fuels-ts/issues/2493

(Back reference added to the issue)
